src/haz3lcore/zipper/Projector.re

```diff
@@ -0,0 +1,315 @@
+  | TextArea => (module Cook(TextAreaCore.M))
+  };
+
+let shape = (p: Map.entry, info: info): shape => {
+  let (module P) = to_module(p.kind);
+  P.placeholder(p.model, info);
+};
+
+/* A projector is replaced by a placeholder in the underlying
+ * editor for view purposes. This projector is an all-whitespace
+ * monotile. Currently there is no explicit notion of placeholders
+ * in the zipper; a tile consisting of any number of whitespaces
+ * is considered a placeholder. This could be made more principled.
+ * Note that a placeholder retains the UUID of the underlying. */
+let placeholder_label = (p: Map.entry, syntax): list(string) =>
+  switch (shape(p, syntax)) {
+  | Inline(width) => [String.make(width, ' ')]
+  | Block({row, col}) => [
+      String.make(row - 1, '\n') ++ String.make(col, ' '),
+    ]
+  };
+
+let placeholder = (p: Map.entry, info: info): syntax =>
+  Piece.Tile({
+    id: Piece.id(info.syntax),
+    label: placeholder_label(p, info),
+    mold: Mold.mk_op(Any, []),
+    shards: [0],
+    children: [],
+  });
+
+/* Must be in-sync with placeholder_label / placeholder above */
+let is_placeholder = (p: Piece.t): bool =>
+  switch (p) {
+  | Tile({label: [s], _}) => s |> String.trim |> String.length == 0
+  | _ => false
+  };
+
+/* Currently projection is limited to convex pieces */
+let minimum_projection_condition = (syntax: syntax): bool =>
+  Piece.is_convex(syntax);
+
+/* Add a new projector, gated on the predicated on the syntax */
+let create = (kind: kind, syntax: syntax): option(Map.entry) => {
+  let (module P) = to_module(kind);
+  P.can_project(syntax) && minimum_projection_condition(syntax)
+    ? Some({kind, model: P.init}) : None;
+};
+
+/* Returns the projector at the caret, if any */
+let indicated = (z: ZipperBase.t) => {
+  open Util.OptUtil.Syntax;
+  let* id = Indicated.index(z);
+  let+ projector = Map.find(id, z.projectors);
+  (id, projector);
+};
+
+/* Is a piece of syntax currently projected? */
+let syntax_is = (ps: Map.t, syntax: option(syntax)): option(Id.t) =>
+  switch (syntax) {
+  | Some(p) when Map.mem(Piece.id(p), ps) =>
+    Map.mem(Piece.id(p), ps) ? Some(Piece.id(p)) : None
+  | _ => None
+  };
+
+/* Is neighboring syntax currently projected? */
+let neighbor_is = (ps, s: Siblings.t): (option(Id.t), option(Id.t)) => (
+  syntax_is(ps, Siblings.left_neighbor(s)),
+  syntax_is(ps, Siblings.right_neighbor(s)),
+);
+
+/* This handles the logic for selecting around projectors.
+ * This amounts to simply selecting over the projected syntax
+ * as if it were atomic instead of selecting into it. In principle,
+ * this should work the same as if the projected syntax was replaced
+ * by its placeholder, the selection was made on that projections,
+ * and then the placeholder was replaced by former syntax. In an
+ * updated syntax model, it should maybe work this way, or else
+ * projected syntax should be represented in the zipper in
+ * a more first-class way */
```

Comment:
I think it would be nice in the future if we moved towards having the projected syntax first-class in the zipper.

Reply:
This may be forced when adding splices, or at least the materialized view / edit state in that case will be sufficiently weighty that it's required for reasonable perf. The complexity here lies in deciding what is the common persistent state versus what is view and/or user dependent, so decisions here are ultimately going to deeply involve both grove and @dm0n3y's syntax model. I've been trying to avoid rewriting the syntax layer but there's definitely some complexity cost to that here

Reply:
Outside the scope of the PR but it is probably useful for us to start having those conversations sooner than later. Even for the view layer I think having a data structure that represents what is viewed fully makes it easier to reason about and test. So we might just need more types.